Anonymous contract parameters generate unfriendly error

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version are you using?

`v20.0.3`

### What did you do?

```rust
#![no_std]
use soroban_sdk::{contract, contractimpl};

#[contract]
pub struct Contract;

#[contractimpl]
impl Contract {
    pub fn empty(_: u64) {}
}
```

```
$ cargo build --target wasm32-unknown-unknown --release
```

### What did you expect to see?

A friendly error message telling me that I can't name parameters on contract functions `_` because that would be a nameless parameter and the contract interface/spec requires names for all parameters.

### What did you see instead?

An error that doesn't tell me where the error is. In this case I could figure out that it was the only parameter on the only function. But in a real contract with multiple functions and multiple parameters would be difficult to debug.

```
$ cargo build --target wasm32-unknown-unknown --release
error: custom attribute panicked
 --> src/lib.rs:7:1
  |
7 | #[contractimpl]
  | ^^^^^^^^^^^^^^^
  |
  = help: message: `""` is not a valid identifier
  = note: this error originates in the attribute macro `contractimpl` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `empty` found for struct `Contract` in the current scope
 --> src/lib.rs:9:12
  |
5 | pub struct Contract;
  | ------------------- function or associated item `empty` not found for this struct
...
9 |     pub fn empty(_: u64) {}
  |            ^^^^^ function or associated item not found in `Contract`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `test_empty` (lib) due to 2 previous errors                                                                                                                                                        
```

--- a/soroban-sdk-macros/src/derive_client.rs
+++ b/soroban-sdk-macros/src/derive_client.rs
@@ -1,6 +1,6 @@
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote};
-use syn::{spanned::Spanned, Error, FnArg, Path, Type, TypePath, TypeReference};
+use syn::{spanned::Spanned, Error, FnArg, Pat, Path, Type, TypePath, TypeReference};
 
 use crate::syn_ext;
 
@@ -172,7 +172,16 @@ pub fn derive_client_impl(crate_path: &Path, name: &str, fns: &[syn_ext::Fn]) ->
                         Ok(ident) => ident,
                         Err(_) => {
                             errors.push(Error::new(t.span(), "argument not supported"));
-                            format_ident!("")
+                            if let FnArg::Typed(pat_type) = t {
+                                if let Pat::Wild(_) = *pat_type.pat.clone() {
+                                    // This will result in a clear compiler error if an underscore is used as an argument.
+                                    format_ident!("_")
+                                } else {
+                                    format_ident!("")
+                                }
+                            } else {
+                                format_ident!("")
+                            }
                         }
                     };
                     (ident, syn_ext::fn_arg_make_ref(t))

--- a/soroban-sdk-macros/src/derive_client.rs
+++ b/soroban-sdk-macros/src/derive_client.rs
@@ -1,6 +1,6 @@
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote};
-use syn::{Error, FnArg, Pat, Path, Type, TypePath, TypeReference};
+use syn::{Error, FnArg, Path, Type, TypePath, TypeReference};
 
 use crate::syn_ext;
 

--- a/soroban-sdk-macros/src/derive_client.rs
+++ b/soroban-sdk-macros/src/derive_client.rs
@@ -172,16 +172,7 @@ pub fn derive_client_impl(crate_path: &Path, name: &str, fns: &[syn_ext::Fn]) ->
                         Ok(ident) => ident,
                         Err(e) => {
                             errors.push(e);
-                            if let FnArg::Typed(pat_type) = t {
-                                if let Pat::Wild(_) = *pat_type.pat.clone() {
-                                    // This will result in a clear compiler error if an underscore is used as an argument.
-                                    format_ident!("_")
-                                } else {
-                                    format_ident!("")
-                                }
-                            } else {
-                                format_ident!("")
-                            }
+                            format_ident!("_")
                         }
                     };
                     (ident, syn_ext::fn_arg_make_ref(t))

--- a/soroban-sdk-macros/src/derive_client.rs
+++ b/soroban-sdk-macros/src/derive_client.rs
@@ -1,6 +1,6 @@
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote};
-use syn::{spanned::Spanned, Error, FnArg, Pat, Path, Type, TypePath, TypeReference};
+use syn::{Error, FnArg, Pat, Path, Type, TypePath, TypeReference};
 
 use crate::syn_ext;
 
@@ -170,8 +170,8 @@ pub fn derive_client_impl(crate_path: &Path, name: &str, fns: &[syn_ext::Fn]) ->
                 .map(|t| {
                     let ident = match syn_ext::fn_arg_ident(t) {
                         Ok(ident) => ident,
-                        Err(_) => {
-                            errors.push(Error::new(t.span(), "argument not supported"));
+                        Err(e) => {
+                            errors.push(e);
                             if let FnArg::Typed(pat_type) = t {
                                 if let Pat::Wild(_) = *pat_type.pat.clone() {
                                     // This will result in a clear compiler error if an underscore is used as an argument.

--- a/soroban-sdk-macros/src/syn_ext.rs
+++ b/soroban-sdk-macros/src/syn_ext.rs
@@ -41,7 +41,10 @@ pub fn fn_arg_ident(arg: &FnArg) -> Result<Ident, Error> {
             return Ok(pat_ident.ident);
         }
     }
-    Err(Error::new(arg.span(), "argument not supported"))
+    Err(Error::new(
+        arg.span(),
+        "argument in this form is not supported, use simple named arguments only",
+    ))
 }
 
 /// Returns a clone of FnArg with the type as a reference if the arg is a typed

